### ISSUE TITLE
Bundling of external deps?

### ISSUE THREAD

**Issue by tunnckoCore** [preconstruct/preconstruct#230]
Sounds and looks good. Read the docs fully.

But basic question about non-making-sense-problem: how, more specifically why it doesn't bundle external deps?

I'm playing on top of the monorepo example `not-random-numbers` where I'm using `the-number-one` inside the `the-number-two`. The general convention is to put it in `devDepdendencies` when you want that dependency **to get bundled with** the source code. Basically everywhere else than `dependencies` and `peerDependencies` because usually such tools are meant to get them and put them in `external` so **won't** get bundled.

But if I put `the-number-one` as devDep, Preconstruct complains that it want it to be on deps or peerDeps.

**Comment by mitchellhamilton** [comment]
I think that allowing bundling of devDeps makes sense. I don't think doing it for devDeps by default makes sense though since most people will generally not want to do it and that would be really confusing(e.g. someone has been using a devDep in some test and they start using it source files but they forget to move it to `dependencies`, it would be really confusing to have it bundled). My thoughts are that dependencies that should be bundled would be specified as `preconstruct#bundledDependencies`(not `bundledDependencies` in the root of the package.json because that has a very specific and different meaning to npm) with any dependencies and transitive dependencies of the bundled dependencies also being bundled along with forcing that bundled deps must be in devDeps and must not be in peerDeps.

```json
{
  "preconstruct": {
    "bundledDependencies": ["some-library"]
  },
  "devDependencies": {
    "some-library": "1.0.0"
  }
}
```

---

> Second. I don't really getting the idea of need for the dev command. I'm getting the idea. But it's unnecessary, since when you are testing your should be using the source file e.g. ../src/index.js in your test/index.js

Importing the source file directly would work for a single package but it breaks when you have transitive internal dependencies, the dev command handles this case. It also lets you require packages directly in node scripts because of the require hook which is especially helpful when building and testing CLIs so you don't have to rebuild it every time you want to run it or use `@babel/register`/`@babel/node`.

> since otherwise you are guaranteed to have lower coverage and not covered pieces because of the nature of compiled code

I don't see how it would affect negatively affect coverage at all, the only thing in the files that the dev command generates is exports of the source files so all of the code in those files will always run.

> Third. I'm guessing what the response for Second will be: "we need it, because while build you will get the-number-one/dist/[name].js is not resolved". True. This can and should be handled by using, as you already is using, the customResolveOptions option to the node-resolve plugin. By passing packageFilter function and overriding (in memory) the main and module fields to point to the src/index.

It's not for the build command at all. We use @rollup/plugin-alias to handle that case though packageFilter is probably a better approach.

**Comment by tunnckoCore** [comment]
> bundled would be specified as preconstruct#bundledDependencies

Agree. I actually don't need bundling of external dependencies anyway. Just was playing, so it's not a rush implementing it. :)

> I don't see how it would affect negatively affect coverage at all, the only thing in the files that the dev command generates is exports of the source files so all of the code in those files will always run.

Ah, good point. True. One little drawback that left is that you need to `../dist/[pkg-name].esm` (if your are using ESModules in tests), which is a bit  not that comfortable, but yea, acceptable.

> We use @rollup/plugin-alias to handle that case

Hm. True, I spot it. But for what I'm seeing it's called only when type is `umd`, so when building cjs & esm it doesn't uses the other package's source but its dist. Which is still not correct?

**Comment by mitchellhamilton** [comment]
> Ah, good point. True. One little drawback that left is that you need to ../dist/[pkg-name].esm (if you're using ESModules in tests), which is a bit not that comfortable (meaning that the user can't/don't know what sanitize/normalization of the pkg.name the preconstruct is doing, if its 1:1 (except the scope) it's okay), but yea, acceptable.

I'm not sure I understand "you need to `../dist/[pkg-name].esm`"? Do you mean import `../dist/[pkg-name].esm`? No one will ever need to import that directly(nor should they), importing by the package name or the directory that the package.json is in will work fine

> Hm. True, I spot it. But for what I'm seeing it's called only when type is umd, so when building cjs & esm it doesn't uses the other package's source but its dist. Which is still not correct?

Other packages will never be bundled(since deps will currently never be bundled except for umd builds) when building cjs and esm so that case will never happen.

**Comment by tunnckoCore** [comment]
> Do you mean import

Yup, exactly.

> package name or the directory that the package.json is in will work fine

Agree.

> Other packages will never be bundled(since deps will currently never be bundled except for umd builds) when building cjs and esm so that case will never happen.

I'm not sure I understand that? :thinking:

**Comment by tunnckoCore** [comment]
But when and if we support bundling of external deps the approach with `plugin-alias` will be a _"MUST CHANGE thing"_ with the `packageFilter` approach. Cuz it won't work quite as expected and well. Meaning, if there is support for bundling dependencies, and one of the packages depend on another from the workspace (as in the case of this issue main topic in the first comment), and using current approach... while `build` is run it will see the `dev`'s dist and bundle that code, instead of the source from `src`.

But anyway, everything is clear.

**Comment by steve-taylor** [comment]
Bundle web apps. Compile libraries with `typescript` or `@babel/cli` directly, but don't bundle them. (They'll get bundled in the web app anyway.) `peerDepdencies` problem solved.

https://www.briefs.fm/3-minutes-with-kent/221
